### PR TITLE
feat(@types/mongodb): better support for Schemas with "any" indexed type

### PR DESCRIPTION
**Pull request by HosseinAgha**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

This PR fixes 2 things:
  - Add better support for Schemas with indexed type set to `any` (tests included)
```typescript
  interface IndexTypeTestModel {
    stringField: string;
    numberField?: number;
    [key: string]: any;
  }
```
  - Change default schema type from `any` into `{ [key: string]: any }` as MongoDB documents can only be objects (we accepted numbers, etc before). Some new tricky bugs in the tests surfaced as the result.

### CODE REVIEW

**Comment by HosseinAgha** [comment]
This can generate quite a lot of complaints from users that relied on the default Schema being `any` and used Polymorphism (__everything extends any__) to change the type of `findX` return value:  

```typescript
// this is OK  as users types is implicitly any
const users = await collection.users.find().toArray();
// this is Wrong as `User` does not extend `Record<string, any>` or `{ [k: string]: any };`
const users: User[] = await collection.users.find().toArray();
// users should do this
const users = await collection.users.find().toArray() as User[];
```

IMHO using `any` for the returned type of `findX` is wrong as we know that it returns an object (`Record<string, any>` to be specific).

**Comment by HosseinAgha** [comment]
__Maintainers I need some other owners' approval before merging this. Please don't merge until someone reviews this__

**Comment by CaselIT** [comment]
does by any change this also fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/39358 ?

**Review by CaselIT — APPROVED** [review]
I agree regarding the change of default.

**Comment by HosseinAgha** [comment]
No I think #39368 is still a TypeScript bug. I opened a proper issue explaining the bug: https://github.com/microsoft/TypeScript/issues/35647

**Comment by HosseinAgha** [comment]
Dear Maintainers, I think this is ready to merge now!

**Comment by CaselIT** [comment]
ok

>  I don't know any workaround to solve it

I also don't. 

Since the typescript issue has been closed, maybe the best way to solve it would be to have an issue here where we explain the problem we are trying to solve and the current implementation and tag one of the typescript devs for an advice?

**Comment by HosseinAgha** [comment]
@CaselIT Good idea. I don't think #39358 have anything to do with this PR. 
I don't think the #39358 is resolvable, I spent some time discussing the issue please take a look at my comment https://github.com/DefinitelyTyped/DefinitelyTyped/issues/39358#issuecomment-566574462 and my issue https://github.com/microsoft/TypeScript/issues/35647#issuecomment-566584170

**Comment by HosseinAgha** [comment]
Regarding this PR, there were some issues after changing DefaultSchema from `any` to `{ [key: string]: any }`. I've added some more test to have more coverage. I'm going to push some fixes today.

**Comment by HosseinAgha** [comment]
I reverted my code to use `any` as `DefaultSchema` I think it generates too much trouble for us.  
I also added some tests for some edge cases. The only scenario that I could not cover is when user provides a Schema with indexed types and a required _id.  
```ts
  interface IndexTypeTestModelWithObjectId {
    _id: ObjectId;
    stringField: string;
    numberField?: number;
    [key: string]: any;
  }
```
I could not make `_id` optional in this scenario (for insertX methods).
I think this is almost impossible to support with the current version of TypeScript.

**Comment by HosseinAgha** [comment]
Anyway, I think we are ready to merge! Thank you for your review @CaselIT.

File: types/mongodb/test/collection/insertX.ts
Comment (CaselIT) [inline types/mongodb/test/collection/insertX.ts]: I think the comment should be `// should not accept wrong _id type`

File: types/mongodb/test/collection/insertX.ts
Comment (CaselIT) [inline types/mongodb/test/collection/insertX.ts]: Do you need to add these tests or they will be added at a later time?

**Comment by CaselIT** [comment]
I've added a couple of minor comments. but they are nice to have :)

File: types/mongodb/test/collection/insertX.ts
Comment (HosseinAgha) [reply types/mongodb/test/collection/insertX.ts]: This is a valid use case but as I said above I couldn't add support for it with the current TS features

**Pull request merged**
